### ISSUE TITLE
Combobox: Handle proper typing

### ISSUE THREAD

**Issue by tskarhed** [grafana/grafana#95386]
At the moment, when enabling custom value, there are type inconsistencies.

https://github.com/grafana/grafana/blob/b1e1297bb358ff2a9fa5cd1abf6c34db15cd8b38/packages/grafana-ui/src/components/Combobox/Combobox.tsx#L168

If you have value types set to numbers and custom value enabled, it will still return a string. [See comment on async PR](https://github.com/grafana/grafana/pull/94147#discussion_r1799413455). We did some initial exploration on fixing this, but put it aside at the time in order to progress.

With this, you can also explore typing of ComboboxOption when returning from the async function, [as seen in this comment](https://github.com/grafana/grafana/pull/94147/files/3714a0ea3e060c44fda9c7c1229373d3019a65ba#r1799451903)

To summarize:

- Proper typing when using custom value
- Proper typing when using the "All" value for MultiCombobox
- Make ComboboxOption["label"] non-optional, if it is in fact returned by the async function
